Update the controller on son_slave connect/disconnect event

morantr opened the pull request:
In the past, we used to have a TCP/IP socket between the Controller and the Agent.
When a radio went down, the son_slave perform reset and closed the socket to the controller.
When that happened, the socket on the controller side was closed too, and the controller removed the node of the son_slave from the database.

Nowadays, we have a BUS infrastructure that replaces the TCP/IP socket. This means that if the son_slave perform reset,
No socket is closed and the controller doesn’t aware that and Agent radio is down.

This PP adds a Topology Query message send after the controller receives M1
and add missing fronthaul radio information to Topology Response message, so the Controller
could remove from its database inactive radio (son_salve) nodes.

#925

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -818,6 +818,13 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_WSC(const std::string &sr
     }
 
     if (!database.setting_certification_mode()) {
+        // trigger Topology query
```
Commit message typo - "upn" --> "upon"

Also, the commit message doesn't explain at all why we send the topology query at this time.
Finally - why do we send it only in certification mode? This will mean it will not work in RDKB and UGW since by default certification mode is not set in these platforms.

morantr replied:
done

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1989,6 +1990,50 @@ bool backhaul_manager::handle_1905_topology_query(ieee1905_1::CmduMessageRx &cmd
     // This is something that cannot be done at this moment because the MediaType field
     // must be computed with information obtained through NL80211_CMD_GET_WIPHY command of DWPAL,
     // which is currently not supported. See "Send standard NL80211 commands using DWPAL #782"
+    // For now, fill from radio data with dummy values based on information we have.
```
It looks like #900 is about to be merged, so I guess it makes more sense to rebase on that instead of adding dummy code. The TODO should mention #900 and not the closed issue in case you insist on not rebasing.

morantr replied:
From what I can see, #900 does not contain relevant code for getting the info we need here.

mariomaz replied:
#900 implements NL80211_CMD_GET_STATION only but NL80211_CMD_GET_WIPHY is required here. Now that the design for the nl80211 standard commands support has been validated, adding the NL80211_CMD_GET_WIPHY command is the next task I'll be working on.

tomereli replied:
Got it - thanks @mariomaz , is there an open issue for the NL80211_CMD_GET_WIPHY support? If not, can you create one?
@morantr please add a comment and a task to replace the dummy values once @mariomaz adds the support.

mariomaz replied:
Created "[TASK] Add NL80211_CMD_GET_WIPHY command #967"

tomereli replied:
> Created "[TASK] Add NL80211_CMD_GET_WIPHY command #967"

Add "Enabler" and "Certification" labels

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1989,6 +1990,50 @@ bool backhaul_manager::handle_1905_topology_query(ieee1905_1::CmduMessageRx &cmd
     // must be computed with information obtained through NL80211_CMD_GET_WIPHY command of DWPAL,
     // which is currently not supported. See "Send standard NL80211 commands using DWPAL #782"
+    // For now, fill from radio data with dummy values based on information we have.
+    for (const auto soc : slaves_sockets) {
```
```suggestion
    for (const auto &soc : slaves_sockets) {
```

morantr replied:
Basically it is a pointer (shared), but you are right, a shared pointer size is 16 and is greater than 4.
fixed.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1989,6 +1990,50 @@ bool backhaul_manager::handle_1905_topology_query(ieee1905_1::CmduMessageRx &cmd
     // which is currently not supported. See "Send standard NL80211 commands using DWPAL #782"
+    // For now, fill from radio data with dummy values based on information we have.
+    for (const auto soc : slaves_sockets) {
+        std::shared_ptr<ieee1905_1::cLocalInterfaceInfo> localInterfaceInfo =
```
```suggestion
        auto localInterfaceInfo =
```

morantr replied:
done

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -276,6 +276,16 @@ bool backhaul_manager::socket_disconnected(Socket *sd)
                 finalize_slaves_connect_state(true, slaves_sockets.front());
             }
             LOG(INFO) << "disconnected slave sockets has been deleted";
+
```
Commit message wrong grammar - "he will be notified" -> "it will be notified"

morantr replied:
done

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2011,47 +2011,61 @@ bool backhaul_manager::handle_1905_topology_query(ieee1905_1::CmduMessageRx &cmd
     // which is currently not supported. See "Send standard NL80211 commands using DWPAL #782"
```
Commit 9fe1245e3381926fb5846b87f4ef0f1951c2d59b should be squashed to commit dedefbf. Also, do not put the commit numbers in the commit message unless you're sure you won't change it again, cause if you force push this commit will be deleted.

morantr replied:
squashed

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1611,6 +1616,83 @@ bool master_thread::handle_cmdu_1905_topology_notification(const std::string &sr
+    // Get all Agent known fronthaul radios if exist, and compare it to what the Agent reports.
+    // If a radio exist on the database but not the tlvDeviceInformation, this radio node needs
+    // to be removed.
+    // This shall work in the opposite way as well: If it is not exist on the database, but reported
```
```suggestion
    // This shall work in the opposite way as well: If it does not exist on the database, but reported
```

morantr replied:
done

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1611,6 +1616,83 @@ bool master_thread::handle_cmdu_1905_topology_notification(const std::string &sr
+    // If a radio exist on the database but not the tlvDeviceInformation, this radio node needs
+    // to be removed.
+    // This shall work in the opposite way as well: If it is not exist on the database, but reported
+    // in the TLV, then needs to add the node to the database. Of course this will remain as
```
```suggestion
    // in the TLV, then we need to add the node to the database. Of course, this will remain as
```

morantr replied:
done

tomereli reviewed (commented):
Mainly nitpick comments, but not approving yet cause I think this should be rebased on #900 instead of using dummy data.

morantr committed:
controller: Send TOPOLOGY_QUERY_MESSAGE upon receiving m1

Send topology message query upon receiving auto-config m1 message.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
tlvf: Change tlv member name

Change the name of the member "ap_channel_band" to "ap_channel_bandwidth".
The term "band" is wrong and not reflects the field's meaning.
Update auto-generated code.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Add media info on radio interfaces in TOPOLOGY RESPONSE

The controller needs to be able to add and remove the radio node from its database based on
the data that is coming from the TOPOLOGY RESPONSE message.
Add dummy data on the wireless fronthaul interfaces in the tlvDeviceInformation which is in
the Topology Response message.
The reason that it is dummy values is that this is something that cannot be done at this moment
because the MediaType field must be computed with information obtained through
NL80211_CMD_GET_WIPHY command of DWPAL, which is currently not supported.
See "Send standard NL80211 commands using DWPAL #782

The filled band data is based on the information we have about the band but it is not accurate.
The fields ap_channel_bandwidth, ap_channel_center_frequency_index1 and
ap_channel_center_frequency_index2 we could get (the bw for sure, the others not so sure),
but it not worth it since it will require to create more internal messaging which will be deleted
after moving to a unified Agent (#435).

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Send a TOPOLOGY NOTIFICATION on son_slave connect/disconnect

When son_slave socket disconnects or connects, send a TOPOLOGY NOTIFICATION with no TLVs to
the controller. The controller will send TOPOLOGY QUERY and on the response, it will be notified
which radio (son_slave) when up or down.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: parse and handle tlvDeviceInformation

Subscribe the controller to TOPOLOGY RESPONSE MESSAGE and
add handler function to it with parsing of device information type TLV.
Add a piece of code that checks whether an unreported fronthaul radio
interface exists on the controller database, and if not, remove its node.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Send TOPOLOGY_QUERY_MESSAGE on empty TOPOLOGY_NOTIFICATION

Send TOPOLOGY_QUERY_MESSAGE when the controller receives TOPOLOGY_NOTIFICATION_MESSAGE
with no Client Association Event TLV. This will result in getting TOPOLOGY_RESPONSE_MESSAGE
with all the information the controller needs to update the topology on its database.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

tomereli reviewed (approved):
Since support for WIPHY commands via NL80211 is not ready yet - approved, but please add a task and a TODO comment in the code and commit message which mention that.

arnout commented on framework/tlvf/yaml/tlvf/ieee_1905_1/s802_11SpecificInformation.yaml:
```diff
@@ -15,7 +15,7 @@ s802_11SpecificInformation:
   network_membership: sMacAddr
   role: eRole
-  ap_channel_band:
+  ap_channel_bandwidth:
```
Yes, I already got confused by that one twice!

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1989,6 +1990,81 @@ bool backhaul_manager::handle_1905_topology_query(ieee1905_1::CmduMessageRx &cmd
+    for (const auto &soc : slaves_sockets) {
+        // Iterate on front radio iface and then switch to back radio iface
+        for (uint8_t i = 0; i < 2; i++) {
+            bool front_iface = bool(i);
```
(nitpick) I find this super ugly. I think it's much better to factor the common part in a function or lambda and call it twice.

morantr replied:
done.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1989,6 +1990,81 @@ bool backhaul_manager::handle_1905_topology_query(ieee1905_1::CmduMessageRx &cmd
+            ieee1905_1::s802_11SpecificInformation media_info = {};
+            localInterfaceInfo->alloc_media_info(sizeof(media_info));
+
+            // BSSID field is not defined well for interface. The common definition is in simple
+            // words "the AP/ETH mac that we are connected to".
+            // For fronthaul radio interface or unused backhaul interface put zero mac.
```
Given that we consider only the radio as an interface, not each VAP, the logical thing for me would be:
- for backhaul STA: the BSSID the STA is associated to (or 0 if not associated)
- for AP: the BSSID of the first VAP, if it is configured. So generally equal to the radio MAC.

morantr replied:
For the backhaul STA, I agree and that what I did.
Regarding the fronthaul interface, I'm not sure this is right. Also, the MAC is already mentioned, so it seems weird to me to write it again on the TLV.
I will keep it as is for now.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1989,6 +1990,81 @@ bool backhaul_manager::handle_1905_topology_query(ieee1905_1::CmduMessageRx &cmd
+            }
+
+            // Note: Not sure that "NON_AP_NON_PCP_STA" is the correct definition to our backhaul
+            // interface.
```
I am sure :-)

morantr replied:
Ok, I'll remove the comment.
The standard defines the following options:

> 
“0100” – non-AP/non-PCP STA
“1000” – Wi-Fi P2P Client (see [B04])
“1001” – Wi-Fi P2P Group Owner (see [B04])
“1010” – 802.11adPCP

Do you know what are the others?
@arnout

arnout replied:
P2P is 11s mesh (AFAIU)

11ad PCP is the mesh-like thing introduced for 60GHz.

Neither of them are covered by MultiAP (AFAIU).

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1989,6 +1990,81 @@ bool backhaul_manager::handle_1905_topology_query(ieee1905_1::CmduMessageRx &cmd
+                front_iface ? ieee1905_1::eRole::AP : ieee1905_1::eRole::NON_AP_NON_PCP_STA;
+
+            // TODO: The Backhaul manager does not hold the information on the front radios.
+            // For now, put zeros and when the Agent management will be move to unified Agent thread
+            // this field will be filled. #435
```
This is not true. The unified agent is about unifying the ap manager and monitor threads. There would still be a separate process for each radio. The idea is that the per-radio process needs to link to bwl and is more prone to crashing, plus it may need more capabilities or different SELinux configuration.

The topology response will always be constructed by the main agent process, not by the per-radio process, so #435 will not solve this.

morantr replied:
Unify Agent is the entity that will connect to two fronthaul managers (per radio) that each one will have a monitor and ap manager. Meaning, it will be the main Agent process.
The backhaul manager will be responsible only to link establishment and remove from him all the Agent stuff.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1989,6 +1990,81 @@ bool backhaul_manager::handle_1905_topology_query(ieee1905_1::CmduMessageRx &cmd
+            media_info.ap_channel_center_frequency_index1 = 0;
+            media_info.ap_channel_center_frequency_index2 = 0;
+
+            auto *media_info_ptr = localInterfaceInfo->media_info(0);
+            if (media_info_ptr == nullptr) {
+                LOG(ERROR) << "media_info is nullptr";
+                return false;
+            }
+
+            std::copy_n(reinterpret_cast<uint8_t *>(&media_info), sizeof(media_info),
+                        media_info_ptr);
```
This is so annoying... There should be a tlvf-generated function that does exactly this (similar to the std::string functions when type is char).

(not for this PR)

morantr replied:
` ¯\_(ツ)_/¯`

tomereli replied:
@LiorAmram Please open a task for this and work on this, you can take a break from the UTAF tasks. Keeping comment open till @LiorAmram sees this.

arnout replied:
#973

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -276,6 +276,16 @@ bool backhaul_manager::socket_disconnected(Socket *sd)
+
+            if (m_eFSMState >= EState::CONNECT_TO_MASTER) {
+                LOG(INFO)
+                    << "Sending topology notification to the controller on son_slave disconnect";
```
Topology notification is multicast, so no "to the controller".

morantr replied:
removed "to the controller".

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -276,6 +276,16 @@ bool backhaul_manager::socket_disconnected(Socket *sd)
+                    LOG(ERROR) << "cmdu creation of type TOPOLOGY_NOTIFICATION_MESSAGE, has failed";
+                    return false;
+                }
+                send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
```
```suggestion
                send_cmdu_to_bus(cmdu_tx, MULTICAST_MAC_ADDR, bridge_info.mac);
```

morantr replied:
Changed to relayed multicast.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -276,6 +276,16 @@ bool backhaul_manager::socket_disconnected(Socket *sd)
             }
             LOG(INFO) << "disconnected slave sockets has been deleted";
+
+            if (m_eFSMState >= EState::CONNECT_TO_MASTER) {
```
Since it's not to controller (see below) it should be any time that the slaves are connected.

morantr replied:
Yes, the 1905 standard does not say it is for the controller but the reason for that is that IEEE 1905.1 does not define entities such as the "Controller" and "Agent". But the whole point of the message is that the controller will receive it.
So there is no point to send it if the controller will not receive it.

arnout replied:
Not necessarily. An agent can also do topology queries. That can be used for keeping track of the backhaul link (#951) or for agent-initiated steering.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1489,6 +1499,15 @@ bool backhaul_manager::handle_slave_backhaul_message(std::shared_ptr<SSlaveSocke
+                return false;
+            }
+            return send_cmdu_to_bus(cmdu_tx, controller_bridge_mac, bridge_info.mac);
+        }
```
Same remarks as above. Maybe move the entire thing into a function (including the state condition).

morantr replied:
Changed to relayed multicast, but I prefer to not turning it to a function since it is not a lot of code and I would like to see the print with the right location.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1611,6 +1616,88 @@ bool master_thread::handle_cmdu_1905_topology_notification(const std::string &sr
+            return false;
+        }
+
+        auto &iface_info = std::get<1>(iface_info_tuple);
```
```suggestion
        const auto &iface_info = std::get<1>(iface_info_tuple);
```

Please please pretty please: do not use strings for MAC addresses.

morantr replied:
It is impossible to add the const.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1611,6 +1616,88 @@ bool master_thread::handle_cmdu_1905_topology_notification(const std::string &sr
+    auto fronthaul_radios_on_db = database.get_node_children(
+        network_utils::mac_to_string(al_mac), beerocks::TYPE_SLAVE, beerocks::STATE_CONNECTED);
+
+    std::unordered_set<std::string> reported_fronthaul_radios;
```
```suggestion
    std::unordered_set<sMacAddress> reported_fronthaul_radios;
```

morantr replied:
I did it because of the database return `set` of `std::string`.
I don't want to change the database in this PR scope.
Let's give #524 more priority.

tomereli replied:
Still, the decision is to use `sMacAddress` instead of `std::string` even if we don't change everything at once, so please change to `sMacAddress` as part of this PR.

arnout replied:
Indeed. It's OK for `fronthaul_radios_on_db` because that is indeed returned by the database. But `reported_fronthaul_radios` gets filled with the MAC addresses that come from the device information so they're not strings. And where you do `reported_fronthaul_radios.find(fronthaul_radio_on_db)` you can easily do `reported_fronthaul_radios.find(network_utils::string_to_mac(fronthaul_radio_on_db))` instead.

Anyway, I'm not going to block the merge for this. But please take it into account next time.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1611,6 +1616,88 @@ bool master_thread::handle_cmdu_1905_topology_notification(const std::string &sr
+            const auto iface_role = media_info->role;
+
+            // For future implementation
+            // const auto &iface_bssid = media_info->network_membership;
+            // const auto iface_bw     = media_info->ap_channel_bandwidth;
+            // const auto iface_cf1    = media_info->ap_channel_center_frequency_index1;
+            // const auto iface_cf2    = media_info->ap_channel_center_frequency_index2;
```
```suggestion
            // In MultiAP context, bssid, bw, cf1 and cf2 are not actually useful since they are more accurately
            // represented in other TLVs (operating channel report, AP operational BSS). 
```

morantr replied:
AFAIR, the current channel selection task does use it. I'm not sure that we will not use it. Keep it for now.

tomereli replied:
Our mindset is to delete what we don't use **now** - so I don't see why we should keep it. In addition, the channel selection task will be stripped down to nothing very soon so we don't really care if it uses it since its obsolete.

arnout replied:
Channel selection task will use this information - but coming from the AP Radio Basic Capabilities and the Channel Preference Report, which is a lot more accurate than the limited information given in this TLV.

I hoped that my suggested edit expressed this feeling accurately...

arnout replied:
After further discusion - it turns out that p_channel_center_frequency_index2 is missing from the Operating Channel Report TLV, so we *will* need to use this for 80+80MHz operation... Which is super-weird...

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1611,6 +1616,88 @@ bool master_thread::handle_cmdu_1905_topology_notification(const std::string &sr
+        // For wireless interface it is defined on IEEE 1905.1 that the size of the media info
+        // is n=10 octets, which the size of s802_11SpecificInformation struct.
+        // For wired interface n=0.
+        if (iface_info.media_info_length() == 10) {
```
I think you should also check the media type, not just the length. It's possible that with some extension, a different media type with a different media info format happens to have the same length.

morantr replied:
According to the standard, only wireless interface has a length of 10. But Ok, I added also an enum check.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -1611,6 +1616,88 @@ bool master_thread::handle_cmdu_1905_topology_notification(const std::string &sr
+    // reported in the TLV, then we need to add the node to the database. Of course this will remain
+    // as "TODO" for future task. Meanwhile parse properly the TLV, leaving the unused parts in
+    // comment for future implementation.
+    auto fronthaul_radios_on_db = database.get_node_children(
```
That's not just fronthaul radios; I think it's really AP radios that you mean (VAPs on that radio can be fronthaul and/or backhaul VAPs; the point it's that it's not backhaul STA).

morantr replied:
No, this tlv refers only to the radio interface and not vaps.
The vaps are handled in AP operational BSS TLV.

arnout replied:
Yes, that's what I mean: there is no such thing as a "fronthaul radio". There are fronthaul and/or backhaul VAPs on a radio, but every radio can be used as fronthaul or backhaul AP.

The only limitation you can have on a radio is that it is not going to be used as an AP.

Since this term "fronthaul radio" is introduced by this PR, I think it's rather important to *not* introduce it if it is not correct (or potentially confusing).

arnout replied:
Actually, the term "fronthaul radio" *is* used all over the place in the test plan, though it is not defined anywhere. But therefore it's OK to use it here as well.

arnout reviewed (approved):
Looks great! I have a few important comments, but everything can be fixed easily, so approve with don't merge.
